### PR TITLE
Move prediction cache to Learner.

### PR DESCRIPTION
**Pull request by trivialfis**
## Clean-ups

* Remove duplicated cache in Learner and GBM.
* Remove ad-hoc fix of invalid cache.
* Remove `PredictFromCache` in predictors.
* Remove prediction cache for linear altogether, as it's only moving the prediction into training process but doesn't provide any actual overall speed gain.
* The cache is now unique to Learner, which means the ownership is no longer shared by any other components.

## Changes

* Add version to prediction cache.
* Use weak ptr to check expired DMatrix.
* Pass shared pointer instead of raw pointer.

## Risks

Before the DMatrix is tied to Booster with a shared pointer, which means the access to it is always safe as it can't expire before the booster.  Now we need to make sure the safety ourselves.

## Advantages

As now we don't store the shared pointer any more,  this should reduce the memory usage when doing grid search with libraries like Scikit-Learn (during evaluation the training matrix can be released).   Also, performance for training continuation (incremental training, checkpoint recovery) will be significantly better as we cache all encountered DMatrix.

Last thing is this PR is intended to ease the development of our DMatrix refactoring plan, which requires an easy to access and reliable prediction cache, so that we can use quantile values as a proxy to actual data.

### PRs:
#5272

### Issues:
#3946
#4786
#4774
#4482

### CODE REVIEW

**Comment by trivialfis** [comment]
Requires some more thoughts to make a clean design for the cache.

**Comment by trivialfis** [comment]
https://github.com/dmlc/xgboost/issues/4482

**Comment by trivialfis** [comment]
@hcho3 We talked about adding a custom segfault handler like: https://github.com/bombela/backward-cpp 

I added a simple one in this PR for debugging on Jenkins (and it was helpful).  But it breaks JVM package's error handling/recovery logic.

File: include/xgboost/predictor.h [hunk 1]
```diff
@@ -32,47 +32,83 @@ namespace xgboost {
...
+   *        DMatrix won't renew it.
+   *
+   *  Passing in a `shared_ptr` is critical here.  First to create a `weak_ptr` inside the
+   *  entry this shared pointer is necessary.  More importantly, the life time the this
+   *  cache is tied to the shared pointer.
+   *
+   *  Another way to make a safe cache is create a proxy to this entry, with anther shared
+   *  pointer defined inside, and pass this proxy around instead of the real entry.  But
+   *  seems to be too messy.  In XGBoost, functions like `UpdateOneIter` will have
+   *  (memory) safe acceess to the DMatrix as long as it's passed in as a `shared_ptr`.
```
Comment (trivialfis) [inline include/xgboost/predictor.h]: Note to myself: `life time the` -> `life time of`, `acceess` -> `access`.

File: include/xgboost/predictor.h [hunk 2]
```diff
@@ -32,47 +32,83 @@ namespace xgboost {
...
+   *               GenericParameter::kCpuId for CPU or positive integer for GPU id.
+   *
+   * \return the cache entry for passed in DMatrix, either an existing cache or newly
+   *         created.
+   */
+  PredictionCacheEntry& Cache(std::shared_ptr<DMatrix> m, int32_t device);
+  /* \brief Get a prediction cache entry.  This entry must be already allocated by `Cache`
+   *        method.  Otherwise a dmlc::Error is thrown.
+   *
+   * \param m pointer the DMatrix.
```
Comment (trivialfis) [inline include/xgboost/predictor.h]: To the

File: src/learner.cc [hunk 3]
```diff
@@ -187,9 +190,13 @@ void GenericParameter::ConfigureGpuId(bool require_gpu) {
...
 class LearnerImpl : public Learner {
  public:
-  explicit LearnerImpl(std::vector<std::shared_ptr<DMatrix> >  cache)
-      : need_configuration_{true}, cache_(std::move(cache)) {
+  explicit LearnerImpl(std::vector<std::shared_ptr<DMatrix> > cache)
+      : need_configuration_{true},
+        cache_{new PredictionContainer} {
     monitor_.Init("Learner");
+    for (std::shared_ptr<DMatrix> const& d : cache) {
+      cache_->Cache(d, GenericParameter::kCpuId);
```
Comment (RAMitchell) [inline src/learner.cc]: So the shared pointer doesn't get copied here?

File: src/learner.cc [hunk 4]
```diff
@@ -741,13 +748,19 @@ class LearnerImpl : public Learner {
...
-      this->PredictRaw(data_sets[i], &preds_[dmat], false);
-      obj_->EvalTransform(&preds_[dmat]);
+      std::shared_ptr<DMatrix> dmat = data_sets[i];
+      auto& predt = this->cache_->Cache(dmat, generic_parameters_.gpu_id);
+      this->ValidateDMatrix(dmat.get());
+      this->PredictRaw(data_sets[i].get(), &predt, false);
+
+      predictions_.SetDevice(generic_parameters_.gpu_id);
+      predictions_.Resize(predt.predictions.Size());
+      predictions_.Copy(predt.predictions);
```
Comment (RAMitchell) [inline src/learner.cc]: Is there an extra copy here that wasn't there before?

File: src/learner.cc [hunk 5]
```diff
@@ -816,14 +829,18 @@ class LearnerImpl : public Learner {
...
     } else if (pred_interactions) {
-      gbm_->PredictInteractionContributions(data, &out_preds->HostVector(), ntree_limit,
+      gbm_->PredictInteractionContributions(data.get(), &out_preds->HostVector(), ntree_limit,
                                             approx_contribs);
     } else if (pred_leaf) {
-      gbm_->PredictLeaf(data, &out_preds->HostVector(), ntree_limit);
+      gbm_->PredictLeaf(data.get(), &out_preds->HostVector(), ntree_limit);
     } else {
-      this->PredictRaw(data, out_preds, training, ntree_limit);
+      auto* predts = &cache_->Cache(data, generic_parameters_.gpu_id);
```
Comment (RAMitchell) [inline src/learner.cc]: What is `predts`? More descriptive name maybe.

File: src/learner.cc [hunk 6]
```diff
@@ -816,14 +829,18 @@ class LearnerImpl : public Learner {
...
+      this->PredictRaw(data.get(), predts, training, ntree_limit);
+      out_preds->SetDevice(generic_parameters_.gpu_id);
+      out_preds->Resize(predts->predictions.Size());
+      out_preds->Copy(predts->predictions);
```
Comment (RAMitchell) [inline src/learner.cc]: Again, is there extra copying happening here?

File: src/learner.cc [hunk 7]
```diff
@@ -965,22 +983,24 @@ class LearnerImpl : public Learner {
...
   HostDeviceVector<GradientPair> gpair_;
   bool need_configuration_;
 
  private:
   /*! \brief random number transformation seed. */
   static int32_t constexpr kRandSeedMagic = 127;
-  // internal cached dmatrix
-  std::vector<std::shared_ptr<DMatrix> > cache_;
+  // internal cached dmatrix for prediction.
+  std::unique_ptr< PredictionContainer > cache_;
```
Comment (RAMitchell) [inline src/learner.cc]: Why is this inside a unique pointer?

File: src/tree/updater_prune.cc [hunk 8]
```diff
@@ -86,9 +87,6 @@ class TreePruner: public TreeUpdater {
         npruned = this->TryPruneLeaf(tree, nid, tree.GetDepth(nid), npruned);
       }
     }
-    LOG(INFO) << "tree pruning end, "
```
Comment (RAMitchell) [inline src/tree/updater_prune.cc]: Why is this removed?

**Comment by trivialfis** [comment]
Wil revert some changes, I sometimes added formatting changes because clangd is warning with clang-tidy checks and it's quite difficult to focus.

File: src/learner.cc (hunk 3 above)
Comment (trivialfis) [reply src/learner.cc]: @RAMitchell We don't keep the shared pointer in booster after this PR.  This way the DMatrix object can be destructed without first deleting the booster.

File: src/learner.cc (hunk 7 above)
Comment (trivialfis) [reply src/learner.cc]: Will replace it with a normal variable.

**Comment by trivialfis** [comment]
> Changing the DMatrix pointer arguments to shared pointers could potentially be extracted.

This might not be possible as we need the shared pointer for constructing `std::weak_ptr`.

**Comment by RAMitchell** [comment]
You could do the shared pointers first. This way we get past some harmless changes before we look at the core logic.

File: src/learner.cc (hunk 4 above)
Comment (trivialfis) [reply src/learner.cc]: @RAMitchell It's a copied moved from `Predictor::PredictFromCache`.  I removed this function altogether as now the cache is managed by `Learner`, hence the copying is also done by `Learner`.  It's 1 copy less as I also removed the temporary vector in C API.

File: src/learner.cc (hunk 5 above)
Comment (trivialfis) [reply src/learner.cc]: Renamed.  But seems to be a convention in XGBoost.

File: src/predictor/cpu_predictor.cc [hunk 9]
```diff
@@ -156,60 +135,78 @@ class CPUPredictor : public Predictor {
...
   void UpdatePredictionCache(
       const gbm::GBTreeModel& model,
       std::vector<std::unique_ptr<TreeUpdater>>* updaters,
-      int num_new_trees) override {
+      int num_new_trees,
+      DMatrix* m,
+      PredictionCacheEntry* predts) override {
     int old_ntree = model.trees.size() - num_new_trees;
     // update cache entry
-    for (auto& kv : (*cache_)) {
```
Comment (trivialfis) [inline src/predictor/cpu_predictor.cc]: It might be worth noting, before this function update the cache for all DMatrix, including validation datasets.  Now it only update the training DMatrix, as I don't think the `updater` can/should have information about validation datasets in the future (otherwise that would be peeking and violates the purpose of validation).

**Comment by trivialfis** [comment]
Not sure why, putting `HostDeviceVector` into thread local store crashes Python on Windows at the end of execution.  @RAMitchell Have you ever experienced this?

**Comment by RAMitchell** [comment]
If you use device vectors as static variables they can get destructed after the cuda context has been released, leading to errors.

**Comment by trivialfis** [comment]
This might make #5207 difficult to implement...

**Comment by trivialfis** [comment]
@RAMitchell Anyway problem for another day.   I cleaned up the PR per your request.  Formatting changes are reverted.  But the change in linear is kept, as gbm no longer has access to global prediction cache, the code removal for linear is necessary.

File: src/predictor/cpu_predictor.cc [hunk 10]
```diff
@@ -156,60 +135,78 @@ class CPUPredictor : public Predictor {
...
     }
-    if (cache_entry->second.predictions.Size() == 0) {
-      // See comment in GPUPredictor::PredictBatch.
-      InitOutPredictions(cache_entry->second.data->Info(),
-                         &(cache_entry->second.predictions), model);
-      cache_entry->second.predictions.Copy(*out_preds);
+    uint32_t const beg_version = predts->version;
+    CHECK_LE(beg_version, end_version);
+
+    if (beg_version < end_version) {
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/predictor/cpu_predictor.cc]: Yes.  So there's no need for predict from cache.

**Comment by trivialfis** [comment]
@RAMitchell 

> If I use two differently sized evaluation matrices, the predictions vector will get resized twice every single iteration. This is a very common case.

Before this PR, Learner stores another prediction vector for each DMatrix (hence 2 prediction vectors for each DMatrix, one for cache one for temporary), I can restore that to avoid the resize.  Another way around is I can modify the input to metric into a `span` and resize the prediction vector when it doesn't have sufficient space for current DMatrix.

**Comment by trivialfis** [comment]
I don't have a better name for `version`.  If you are boosting forest with 4 trees per-forest, the updated version each round is 4.  This is non-ideal, but right now there's not much we can do.

As another problem regarding refresh updater I mentioned offline with @RAMitchell , it should be fine as training continuation requires at least 1 serialization.  The cache is cleared.

**Comment by mli** [comment]
# [Codecov](https://codecov.io/gh/dmlc/xgboost/pull/5220?src=pr&el=h1) Report
> Merging [#5220](https://codecov.io/gh/dmlc/xgboost/pull/5220?src=pr&el=desc) into [master](https://codecov.io/gh/dmlc/xgboost/commit/29eeea709aff275fa97187858906fce711b4ae34&el=desc) will **not change** coverage by `%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/dmlc/xgboost/pull/5220?src=pr&el=tree)

```diff
@@           Coverage Diff           @@
##           master    #5220   +/-   ##
=======================================
  Coverage   83.76%   83.76%           
=======================================
  Files          11       11           
  Lines        2409     2409           
=======================================
  Hits         2018     2018           
  Misses        391      391           
```



------

[Continue to review full report at Codecov](https://codecov.io/gh/dmlc/xgboost/pull/5220?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/dmlc/xgboost/pull/5220?src=pr&el=footer). Last update [29eeea7...4fcc43e](https://codecov.io/gh/dmlc/xgboost/pull/5220?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by trivialfis** [comment]
The copying is annoying.  I don't have any thoughts on to get rid of them yet.
